### ISSUE TITLE
Review API around store_unary, store_full_arg etc

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#2132]
#2130 added the ``store_unary`` option as a quick way of getting the basic changes in to allow us track these extra nodes, but we need to carefully look at this before release to think about what we want the long-term API to be and whether we want to deprecate the ``store_full_arg`` option in favour of something more flexible.

**Comment by jeromekelleher** [comment]
In discussion with @GertjanBisschop this morning we decided that one way to do it would be to decompose the functionality in terms of how we create new nodes and edges, and have some rules for doing this. Specifically we have two new arguments call (something like) ``nodes`` and ``edges``. These have the possible values:

**nodes**
- COALESCENT
- COMMON_ANCESTOR
- RECOMBINANT
- MIGRANT
- PASS_THROUGH

**edges**
- COALESCING_SEGMENTS
- EVERYWHERE

The current default would be ``nodes=COALESCENT`` and ``edges=COALESCING_SEGMENTS``, and the ``full_arg`` option would correspond to ``nodes=COMMON_ANCESTOR|RECOMBINANT|MIGRANT`` and ``edges=EVERYWHERE``. The functionality we're looking for in the pedigree simulation (#2129) would be ``nodes=PASS_THROUGH`` and ``edges=EVERYWHERE``.

This should be reasonably flexible and implementable across the different models. Any thoughts? Ping @sgravel @petrelharp @hyanwong

**Comment by petrelharp** [comment]
This sounds like a very nice proposal, although it doesn't quite translate to the python api?

**Comment by jeromekelleher** [comment]
Talking to @GertjanBisschop again about this we've decided that really we just have one interesting parameter, ``additional_nodes`` which defines when we create a new node associated with a particular ancestral haplotype, besides the ones we must create to record coalescences. The second parameter is a boolean (say) ``coalescing_segments_only`` which controls whether we keep the current (odd, in retrospect) behaviour of only recording relationships between segments that have coalesced. This can only be used *without* the ``additional_nodes`` argument, because it doesn't make sense to record (say) migrant nodes without keeping track of how they relate to other nodes.

We could make the ``additional_nodes`` argument take an [enum](https://docs.python.org/3.7/library/enum.html) which might look something like this:
```python
import enum

class NodeType(enum.Flag):
        COMMON_ANCESTOR = 1
        RECOMBINANT = 2
        MIGRANT = 4
        PASS_THROUGH = 8
```

So, if you wanted to capture all recombinant and migrant nodes as well as the coalescent ones (but not common ancestor) nodes we'd do:

```python
ts = msprime.sim_ancestry(..., 
             additional_nodes=msprime.NodeType.RECOMBINANT  | msprime.NodeType.MIGRANT)
```

Note that specifying any ``additional_nodes`` means that we always record the full span of relationships between these nodes.

**Comment by jeromekelleher** [comment]
Note that this means the ``additional_nodes`` and ``coalescing_segments_only`` are therefore more-or-less independent, and that ``coalescing_segments_only`` is essentially what has been implemented in #2130.

There's some discussion to be had around how these parameters would interact with the ``fixed_pedigree'' model, but I think we can vary the defaults a bit across models without making things too confusing.

**Comment by petrelharp** [comment]
This generally sounds good; I don't have the details of what exactly the options above do in my head, but would be happy to talk through them (or read a proposed docstring) if you want more input.

**Comment by jeromekelleher** [comment]
OK great, sounds like we have a plan here then.

**Comment by GertjanBisschop** [comment]
Implemented in #2162
